tuf/ngclient/updater.py

```diff
@@ -316,76 +319,63 @@ def _load_targets(self, role: str, parent_role: str) -> None:
-
+        targetinfo = None
+        terminated = False
         # Preorder depth-first traversal of the graph of target delegations.
-        while number_of_delegations > 0 and len(role_names) > 0:
-
-            # Pop the role name from the top of the stack.
-            role_name, parent_role = role_names.pop(-1)
-
-            # Skip any visited current role to prevent cycles.
-            if (role_name, parent_role) in visited_role_names:
-                logger.debug("Skipping visited current role %s", role_name)
-                continue
-
-            # The metadata for 'role_name' must be downloaded/updated before
-            # its targets, delegations, and child roles can be inspected.
-            self._load_targets(role_name, parent_role)
-
-            role_metadata: Targets = self._trusted_set[role_name].signed
-            target = role_metadata.targets.get(target_filepath)
-
-            if target is not None:
-                logger.debug("Found target in current role %s", role_name)
-                return {"filepath": target_filepath, "fileinfo": target}
-
-            # After preorder check, add current role to set of visited roles.
-            visited_role_names.add((role_name, parent_role))
-
-            # And also decrement number of visited roles.
-            number_of_delegations -= 1
-
-            if role_metadata.delegations is not None:
-                child_roles_to_visit = []
-                # NOTE: This may be a slow operation if there are many
-                # delegated roles.
-                for child_role in role_metadata.delegations.roles:
-                    if child_role.is_in_trusted_paths(target_filepath):
-                        logger.debug("Adding child role %s", child_role.name)
-
-                        child_roles_to_visit.append(
-                            (child_role.name, role_name)
-                        )
-                        if child_role.terminating:
-                            logger.debug("Not backtracking to other roles.")
-                            role_names = []
-                            break
-                # Push 'child_roles_to_visit' in reverse order of appearance
-                # onto 'role_names'.  Roles are popped from the end of
-                # the 'role_names' list.
-                child_roles_to_visit.reverse()
-                role_names.extend(child_roles_to_visit)
-
-        if number_of_delegations == 0 and len(role_names) > 0:
-            logger.debug(
-                "%d roles left to visit, but allowed to visit at most %d.",
-                len(role_names),
-                self.config.max_delegations,
-            )
-
-        # If this point is reached then target is not found, return None
-        return None
+        if number_of_delegations <= 0:
+            return targetinfo, terminated
+
+        # Pop the role name from the top of the stack.
+        role_name, parent_role = current_role_pair
```

Comment:
From the type annotation of `current_role_pair` I see it's a `List[Tuple[str, ...]`.
Don't you need to get an exact element from `current_role_pair` which will contain `role_name` and `parent_role` pair?